### ISSUE TITLE
xct_core is suddenly invalidated (all fields zeroed-out)

### ISSUE THREAD

**Issue by caetanosauer** [caetanosauer/zero#12]
In a long-running benchmark, some assertions that verify the state of the transaction object -- particularly the "core" object `xct_core` -- start failing without any apparent reason. Further inspection shows that when the assertions fail, all fields in the core object are set to zero, as if all its binary contents were updated with memset(.., 0, ...).

Some of the failing assertions are:

``` c++
 w_assert1(_core == NULL || _tid == _core->_tid);
```

on xct_t::tid

``` c++
w_assert3(state() == xct_active);
```

on xct_t::xct_t (constructor)

Curiously, the bug only seems to happen when the buffer pool is small enough so that replacements are frequent. I could not reproduce it with a large buffer. Maybe it's related to one of the cleaner/eviction issues.

**Comment by caetanosauer** [comment]
The assertion on `tid()` seems to occur only when called from `xct_t::_flush_logbuf`

**Comment by caetanosauer** [comment]
The bug seems to happen in some places where a new logrec_t is allocated/constructed.

For example, in the log stubs (logstub_gen.cpp);

``` c++
W_DO(xd->get_logbuf(logrec, t_btree_overwrite));
new (logrec) btree_overwrite_log(page, key, old_el, new_el, offset, elen);
W_DO(xd->give_logbuf(logrec, &page)); 
```

Before calling `give_logbuf`, xct_core is already invalid.

Another example is in the constructor of xct_t:

``` c++
_log_buf = new logrec_t;
_log_buf_for_piggybacked_ssx = new logrec_t;
```

The invalidation occurs after either the first or the second invocation to `new logrec_t`

**Comment by caetanosauer** [comment]
The bug is being caused by the new sysevent log records, being invoked, for instance, on page writes. The code looks like this:

``` c++
void sysevent::log_page_write(shpid_t shpid, uint32_t count)     
{
    logrec_t* lr = new logrec_t();
    lr->header._type = logrec_t::t_page_write;
    lr->header._cat = 0 | logrec_t::t_status;

    memcpy(lr->_data, &shpid, sizeof(shpid_t));
    memcpy(lr->_data + sizeof(shpid_t), &count, sizeof(uint32_t));
    lr->fill(0, sizeof(shpid_t) + sizeof(uint32_t));
    W_COERCE(smlevel_0::log->insert(*lr, NULL));
}
```

No idea why this would cause the xct_core object to become corrupt. Investigating it further for now.

**Comment by caetanosauer** [comment]
The problem was actually a (pretty obvious) memory leak. The sysevent methods were allocating a `logrec_t` object with `new` but not releasing it with `delete`. It's solved now.

Fixed on d6affb1

However, it's still weird that the TLS allocator (block_alloc) would just start corrupting memory instead of throwing some kind of exception. I'm opening another issue for that.
